### PR TITLE
Added mobile auth support + trade confirmations

### PR DESCRIPTION
**Pull request by waylaidwanderer**
Features:
* mobile auth support
* trade offer confirmations
* live trade confirmations

AppVeyor will likely fail to build this after merging. Needs some pre-build commands added in "/settings/build" URL:
* Set "Visual Studio solution or project file" to `SteamBot.sln`
* In "Before build script", add
    `git submodule update --init --recursive` and `nuget restore SteamAuth\SteamAuth\SteamAuth.sln`

Have a look and feel free to test (confirmed working on my [SteamTradeOffersBot](https://github.com/waylaidwanderer/SteamTradeOffersBot/) fork).

### CODE REVIEW

**Comment by geel9** [comment]
I don't think that linking an account should be part of SteamBot. You should use a tool like the Steam Desktop Authenticator and then import the maFile.

**Comment by waylaidwanderer** [comment]
Why force the user to download two separate things when they can just use SteamBot alone? The auth files are saved to the `authfiles` folder, named `{username}.authfile`, which has exactly the same content as `maFile`.

**Comment by geel9** [comment]
Because it's out of scope. SteamBot doesn't register accounts for you either.

**Comment by waylaidwanderer** [comment]
That sounds like it'd be against TOS to automate it, but this isn't the same thing. SteamAuth allows for delinking IIRC, so it shouldn't be a problem if the user doesn't want to use SteamBot for authing his bot account anymore. 

I suppose we should get some more opinions on this issue. I think it's more useful/convenient this way, but I'm not against removing the function.

**Comment by geel9** [comment]
On trade confirmations you need to handle the `WGTokenInvalidException` from the latest version of SteamAuth. Ideally you'd set the `SteamGuardAccount.Session.SteamLogin` and `SteamGuardAccount.Session.SteamLoginSecure` properties to the bot's own SteamLogin and SteamLoginSecure cookies before doing confirmation-related tasks.

**Comment by waylaidwanderer** [comment]
Ok, I'll undo the last commit and redo it.

**Comment by waylaidwanderer** [comment]
That's everything, I think. Mobile auth support, trade offer confirmations, live trade confirmations. Ready for testing/merging, pending further feedback.

File: SteamBot/Bot.cs [hunk 1]
```diff
@@ -336,6 +364,24 @@ public void HandleBotCommand(string command)
...
+            consoleInput = null;
+            while (true)
+            {
+                if (consoleInput != null)
+                {
+                    return consoleInput;
+                }
+                Thread.Sleep(5);
+            }
+        }
```
Comment (BlueRaja) [inline SteamBot/Bot.cs]: ಠ_ಠ

Comment (Bottswana) [inline SteamBot/Bot.cs]: Uuurm, yeah.
*internal monologue* aaaaaaaaarrrgggh

Comment (waylaidwanderer) [inline SteamBot/Bot.cs]: No different than this: https://github.com/Jessecar96/SteamBot/blob/master/SteamBot/Bot.cs#L942

**Comment by Jessecar96** [comment]
SteamBot should definitely have the ability to handle linking mobile authenticators.
As soon as someone else approves of this it needs to be merged asap.

File: SteamBot/Bot.cs (hunk 1 above)
Comment (Bottswana) [inline SteamBot/Bot.cs]: We should probably switch to using ResetEvents wait intervals than
Thread.Sleep should we not?

Comment (waylaidwanderer) [inline SteamBot/Bot.cs]: Feel free to make a pull request later if you feel it's necessary, but I think it's important that we get mobile auth working with SteamBot ASAP first.

Comment (Bottswana) [inline SteamBot/Bot.cs]: Alright, we can review that later and possibly the other place it occurs in
the code base after, if its that  important.

**Comment by BlueRaja** [comment]
Code looks fine except for that weird bit, which given the severity of this issue is not important.

However I will not be able to test this before the end of the week.  Could someone else?

**Comment by BlueRaja** [comment]
Code looks fine except for that weird bit, which given the severity of this issue is not important.

_(also I'm not a fan of the submodule, but we can fix that later.  I'll have to talk to @geel9 about pushing SteamAuth to Nuget)_

However I will not be able to test this before the end of the week.  Could someone else?

**Comment by waylaidwanderer** [comment]
I just tested this on linux using mono, used the command `exec [index] linkauth` and got an error about some mono Crypto library before the bot manager crashed. Possibly a problem with SteamAuth. Can someone reproduce this?

**Comment by NoChanceSD** [comment]
Tried to use this but after typing the input command with the steam guard code, just nothing happened.
Not sure what's wrong here.

**Comment by waylaidwanderer** [comment]
Index starts at 0 for your first bot.

**Comment by NoChanceSD** [comment]
I'm aware of that, your method SendInput seems to have 0 references, maybe that's the issue?

**Comment by waylaidwanderer** [comment]
My bad, I missed that when I was copy-pasting stuff over. Just pushed a commit to fix it.

File: SteamBot/Bot.cs (hunk 1 above)
Comment (Lagg) [inline SteamBot/Bot.cs]: Hey, donchu bring me into this @waylaidwanderer . I wrote it that way because I needed to get rid of the winform dependency in half an hour and I think we can agree sir that a GUI is worse than a sleeping thread that can't do anything useful anyway :(

Then again, I guess that's a perfectly valid argument for you too. Hm. Still though it's good people are paying attention to what really matters about this PR.

**Comment by bartico6** [comment]
Don't you think using "ConfirmAllMobileConfirmations" is a bit unsafe? I have no idea how the confirmation ID is linked to the trade id... Unable to automate imo :/

**Comment by NoChanceSD** [comment]
Alright, now everything seems to be working fine. It's good to be merged in my opinion.

**Comment by hydracore** [comment]
I've took the liberty of testing this pull request, and this is the only result it gives me:
```
[Test Bot 2015-12-02 23:13:01] INFO: Connecting...
[Test Bot 2015-12-02 23:13:01] SUCCESS: Done Loading Bot!
[Test Bot 2015-12-02 23:13:03] ERROR: System.AggregateException: One or more errors occurred. ---> System.ArgumentNullException: String reference not set to an instance of a String.
Parameter name: s
   at System.Text.Encoding.GetBytes(String s)
   at SteamKit2.WindowsInfoProvider.GetDiskId()
   at SteamKit2.HardwareUtils.GenerateMachineID()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(TimeSpan timeout)
   at SteamKit2.HardwareUtils.GetMachineID()
   at SteamKit2.SteamUser.LogOn(LogOnDetails details)
   at SteamBot.Bot.UserLogOn() in c:\Users\Administrator\Documents\Visual Studio 2012\2FA_TEST\SteamBot\SteamBot\Bot.cs:line 845
   at SteamBot.Bot.<HandleSteamMessage>b__3(ConnectedCallback callback) in c:\Users\Administrator\Documents\Visual Studio 2012\2FA_TEST\SteamBot\SteamBot\Bot.cs:line 456
   at SteamKit2.CallbackMsgExtensions.Handle[T](ICallbackMsg msg, Action`1 handler)
   at SteamBot.Bot.HandleSteamMessage(ICallbackMsg msg) in c:\Users\Administrator\Documents\Visual Studio 2012\2FA_TEST\SteamBot\SteamBot\Bot.cs:line 450
   at SteamBot.Bot.BackgroundWorkerOnDoWork(Object sender, DoWorkEventArgs doWorkEventArgs) in c:\Users\Administrator\Documents\Visual Studio 2012\2FA_TEST\SteamBot\SteamBot\Bot.cs:line 1102
---> (Inner Exception #0) System.ArgumentNullException: String reference not set to an instance of a String.
Parameter name: s
   at System.Text.Encoding.GetBytes(String s)
   at SteamKit2.WindowsInfoProvider.GetDiskId()
   at SteamKit2.HardwareUtils.GenerateMachineID()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()<---
 
[Test Bot 2015-12-02 23:13:03] WARN: Restarting bot...
```
Tested on a clean install so to say. Downloaded the repo just for this purpose. Works fine without this pull request.

**Comment by WildCard65** [comment]
@hydracore Your issue is this: https://github.com/SteamRE/SteamKit/issues/200

**Comment by GreYzZ** [comment]
@waylaidwanderer 
why is the steam auth folder empty if I download the zip?

**Comment by RafalNiewinski** [comment]
@GreYzZ 
Because it is attached as git submodule and should be downloaded separately.
However, I recommend you use git.

**Comment by GreYzZ** [comment]
@RafalNiewinski 
can you explain how to do this, sorry I'm not very used on github.
Or can you send me the bot with submodule integrated?

**Comment by AleksaSavic3** [comment]
1) Download waylaid's Steam Bot repository and geel9's SteamAuth. Extract SteamAuth to the appropriate folder.
2) Learn how Git works, download Github for desktop and click the button near an URL that GitHub gives you on repository home page. Github app will open, ask you where to clone the repository and you can enjoy Steam Bot!
3) Use git via command line.

Which is easier to do is up to you. :+1:

**Comment by Jessecar96** [comment]
Merging, everything seems alright here.

**Comment by doktokto** [comment]
I just getting this error. What's wrong?
I am downloaded using GitHub and just changed login, password, apikey and name.
fortime

**Comment by doktokto** [comment]
Okay, I used ```exec 0 linkauth```, first time I didn't received sms so I'm just closed the bot and started again. I received login error "Invalid password". I thought the reason was unsuccessful "linkauth" command. I'm wrote again ```exec 0 linkauth``` and received e-mail, sms, input them to console using commands and all was okey, I received green color message and I'm closed bot. I'm started bot again and received error "AccountLogonDeniedNeedTwoFactorAuth" about 5 times and then again InvalidPassword. Why this happened?

**Comment by waylaidwanderer** [comment]
With version 1.6.5 of Git and later, you can use:

    git clone --recursive git://github.com/foo/bar.git
    cd bar

For already cloned repos, or older Git versions, just use:

    git clone git://github.com/foo/bar.git
    cd bar
    git submodule update --init --recursive

From [StackOverflow](http://stackoverflow.com/questions/3796927/how-to-git-clone-including-submodules)

This will download SteamAuth as a submodule.

**Comment by doktokto** [comment]
Maybe will be better to start mobile auth when getting error about ```TwoAuthCode```, not after ```exec [index] linkauth```?

**Comment by waylaidwanderer** [comment]
See https://github.com/Jessecar96/SteamBot/pull/855.

@doktokto If you run into `EResult.AccountLogonDeniedNeedTwoFactorCode`, it means 2FA is already enabled on that account, so you wouldn't be linking an authenticator in that case.

**Comment by BestPotatoEU** [comment]
After I clone SteamBot and build the solution I try to run it but I get this error:
Error	Metadata file ...\SteamBot\SteamAuth\SteamAuth\bin\Debug\SteamAuth.dll' could not be found

**Comment by xxbone** [comment]
@waylaidwanderer I get a generalfailure error after I add my phone number. Am I doing something wrong? Ive tried it over 2-3 days multiple times always get the same error.

**Comment by waylaidwanderer** [comment]
@xxbone You might need to figure out what's causing it. Do a Console.WriteLine before each "return LinkResult.GeneralFailure" in SteamAuth's function and see which one it is.

**Comment by geel9** [comment]
Or just..debug it?

**Comment by xxbone** [comment]
Got it working @waylaidwanderer 
I had one more doubt, if I want to use my main acc as a bot, I'd have to link an authenticator to it via steambot, that would lead me to not be able to view the 2FA codes on my mobile. So if I have to do manual trades or logins, how do I go around doing that?

**Comment by geel9** [comment]
@xxbone use the [Steam Desktop Authenticator](https://github.com/Jessecar96/SteamDesktopAuthenticator)

**Comment by xxbone** [comment]
Ah great thanks @geel9

**Comment by waylaidwanderer** [comment]
@geel9 I find this way easier for people to understand. I know I got really confused when I was trying to figure out how to debug for the first time :P

**Comment by waylaidwanderer** [comment]
That said, I'm getting this error when adding my phone number:

`{"success":false,"error_text":"Sorry, we couldn't send an SMS to your phone","fatal":true}`

It worked for 6 bots in a row and then stopped working for the last 3. There's no limit on how many accounts you can use with 1 number, right? Do you know what the problem is? Should I just try again later?

**Comment by xxbone** [comment]
@geel9 using the SDA, after logging in nothing happens. Like literally nothing happens. I've cross checked my password again and again, I don't get a SMS code and the app is doing nothing.

**Comment by geel9** [comment]
@waylaidwanderer there's a limit on how many accounts you can add a phone number to within a certain time period (probably 24hr)

@xxbone if the account you're trying to log into already has the authenticator on it, you'll need to copy the authenticator file into the maFiles directory. You'll also need to do some editing of the manifest.json file. If possible, the easiest way to do it all is to unlink the authenticator, add it using the SDA, then copy and rename the .maFile that's generated into your SteamBot directory.

**Comment by xxbone** [comment]
@geel9  I did unlink auth from it before adding it to SDA. What im thinking is that it's being caused because of the InvalidPassword error I was getting before after I linked and unlinked auth from the bot and added it to my mobile. I guess I should wait a couple hours?

**Comment by waylaidwanderer** [comment]
@geel9 Thanks, I'll try again tomorrow.

**Comment by geel9** [comment]
@xxbone I'm not sure why the InvalidPassword error occurs but it seems to be some sort of rate-limiting.

**Comment by scholtzm** [comment]
Yep, they also updated protobufs for this. See the SteamKit repo.

**Comment by xxbone** [comment]
Yeah thought so @geel9  Thanks for all the help.
You too @waylaidwanderer

**Comment by adlan05** [comment]
Guys can someone make a video on how to compile it and run it please.

**Comment by AleksaSavic3** [comment]
https://www.youtube.com/watch?v=7UiaPZNwqsw

TL;DR: Learn to code.
